# Data Cleaning with Pandas

In [185]:
import pandas as pd

In [3]:
"""medium data: excel can't open, tableau can"""

"medium data: excel can't open, tableau can"

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [4]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')

/Users/krex/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [5]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [6]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [7]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.5+ MB


In [8]:
import pandas as pd
import numpy as np
import warnings
np.seterr(all='warn')
warnings.simplefilter("always")

In [9]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/Users/krex/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [10]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [11]:
bldg_df.head().T

0                            1  \
Major                                    4300                         4610   
Minor                                     167                          399   
BldgNbr                                     1                            1   
NbrLivingUnits                              1                            1   
Address             15223   40TH AVE S  98188  4431   FERNCROFT RD   98040   
BuildingNumber                          15223                        4431    
Fraction                                                                     
DirectionPrefix                                                              
StreetName          40TH                         FERNCROFT                   
StreetType                             AVE                          RD       
DirectionSuffix                            S                                 
ZipCode                                 98188                        98040   
Stories                                     1                            2   
BldgGrade                                   6                            9   
BldgGradeVar                                0                            0   
SqFt1stFloor                              990                          980   
SqFtHalfFloor                               0                            0   
SqFt2ndFloor                                0                         1140   
SqFtUpperFloor                              0                            0   
SqFtUnfinFull                               0                            0   
SqFtUnfinHalf                               0                            0   
SqFtTotLiving                             990                         2950   
SqFtTotBasement                           990                          940   
SqFtFinBasement                             0                          830   
FinBasementGrade                            0                            9   
SqFtGarageBasement                          0                            0   
SqFtGarageAttached                          0                          480   
DaylightBasement                                                         Y   
SqFtOpenPorch                              80                            0   
SqFtEnclosedPorch                           0                            0   
SqFtDeck                                    0                          600   
HeatSystem                                  6                            5   
HeatSource                                  1                            2   
BrickStone                                  0                            0   
ViewUtilization                                                          N   
Bedrooms                                    2                            3   
BathHalfCount                               0                            1   
Bath3qtrCount                               0                            0   
BathFullCount                               1                            2   
FpSingleStory                               0                            0   
FpMultiStory                                0                            0   
FpFreestanding                              0                            2   
FpAdditional                                0                            0   
YrBuilt                                  1922                         1978   
YrRenovated                                 0                            0   
PcntComplete                                0                            0   
Obsolescence                                0                            0   
PcntNetCondition                            0                            0   
Condition                                   3                            4   
AddnlCost                                   0                            0   

                                              2                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [12]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [13]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511412 entries, 0 to 511411
Data columns (total 4 columns):
Major            511412 non-null int64
Minor            511412 non-null int64
SqFtTotLiving    511412 non-null int64
ZipCode          468372 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [14]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [15]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'])

ValueError: Unable to parse string "      " at position 935519

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [16]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [17]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')



Did it work?

In [18]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.5+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [19]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [20]:
sales_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.5+ MB


Now, let's try our join again.

In [54]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [55]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [56]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1437522 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1437522 non-null float64
Minor            1437522 non-null float64
DocumentDate     1437522 non-null object
SalePrice        1437522 non-null int64
SqFtTotLiving    1437522 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 76.8+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [57]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
99   858140.0  376.0   05/22/2012          0            900     NaN
100  858140.0  376.0   11/28/2017          0            900     NaN
152  720319.0  520.0   11/20/2013     699950           2840     NaN
153  720319.0  520.0   09/16/2013          0           2840     NaN
163  894677.0  240.0   12/21/2016     818161           2450     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [58]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]
#creats a boolean mask, 1 for zip codes, 0 for not
#~ flips a boolean , ~salesdata in this case makes zeros for values for zip codes, and 1 for no zip code
sales_data.tail()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
1437517  947794.0  320.0   02/22/2002     331681           3010   98059
1437518  404080.0   40.0   05/31/2018          0           1060   98008
1437519  721470.0  820.0   04/15/2005     484950           2620   98077
1437520  235430.0  782.0   06/01/1984      85000           1500   98027
1437521  401990.0  175.0   05/30/2006          0           1840   98155

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [219]:
sales_price = sales_data.loc[sales_data['SalePrice'] > 0, :]
sales_price.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 867413 entries, 0 to 1437520
Data columns (total 6 columns):
Major            867413 non-null float64
Minor            867413 non-null float64
DocumentDate     867413 non-null object
SalePrice        867413 non-null int64
SqFtTotLiving    867413 non-null int64
ZipCode          867413 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 46.3+ MB


In [215]:
sales_price = sales_price.loc[sales_price['SqFtTotLiving'] > 0, :]

In [220]:
sales_price.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 867413 entries, 0 to 1437520
Data columns (total 6 columns):
Major            867413 non-null float64
Minor            867413 non-null float64
DocumentDate     867413 non-null object
SalePrice        867413 non-null int64
SqFtTotLiving    867413 non-null int64
ZipCode          867413 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 46.3+ MB


### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [221]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

sales_price.loc[sales_price['ZipCode'].apply(is_integer) == False, 'ZipCode'].head()

10168    98042-3001
10169    98042-3001
10170    98042-3001
10171    98042-3001
16528    98199-3014
Name: ZipCode, dtype: object

In [223]:
sales_price['ZipCode'] = sales_price['ZipCode'].apply(lambda x: str(x)[:5])
sales_price.info()

/Users/krex/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


<class 'pandas.core.frame.DataFrame'>
Int64Index: 867413 entries, 0 to 1437520
Data columns (total 6 columns):
Major            867413 non-null float64
Minor            867413 non-null float64
DocumentDate     867413 non-null object
SalePrice        867413 non-null int64
SqFtTotLiving    867413 non-null int64
ZipCode          867413 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 46.3+ MB


In [165]:
#def first_five(x):
   # y = str(x)
  #  z = y[:5]
   # return z

In [181]:
#sales_price.loc[sales_price['ZipCode'].apply(is_integer) == False, 'ZipCode'].tail()

1437398    98034.
1437401    98198.
1437467    98168.
1437502    98115.
1437519    98077.
Name: ZipCode, dtype: object

In [212]:
#sales_price['ZipCode'] = sales_price['ZipCode'].apply(lambda x: str(x)[:6])
#sales_price['ZipCode'] = sales_price['ZipCode'].apply(lambda x: x.strip)
#sales_price['ZipCode'] = sales_price['ZipCode'].apply(lambda x: int(x))


TypeError: int() argument must be a string, a bytes-like object or a number, not 'builtin_function_or_method'

In [35]:
#sales_zips = sales_prices_2.loc[~sales_prices_2['ZipCode'].isna(), :]
#sales_zips

NameError: name 'sales_prices_2' is not defined

### 3. Add a column for PricePerSqFt



In [230]:
sales_price['PricePerSqFt'] = sales_price['SalePrice'] / sales_price['SqFtTotLiving']
sales_price.head()

/Users/krex/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0  138860.0  110.0   08/21/2014     245000           1490   98002   
1  138860.0  110.0   06/12/1989     109300           1490   98002   
2  138860.0  110.0   01/16/2005      14684           1490   98002   
6  423943.0   50.0   07/15/1999      96000            960   98092   
7  423943.0   50.0   01/08/2001     127500            960   98092   

   PricePerSqFt  
0    164.429530  
1     73.355705  
2      9.855034  
6    100.000000  
7    132.812500

In [229]:
sales_price['PricePerSqFt'] = sales_price["SalePrice"]/sales_price["SqFtTotLiving"]
sales_price.info()

/Users/krex/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


<class 'pandas.core.frame.DataFrame'>
Int64Index: 867413 entries, 0 to 1437520
Data columns (total 7 columns):
Major            867413 non-null float64
Minor            867413 non-null float64
DocumentDate     867413 non-null object
SalePrice        867413 non-null int64
SqFtTotLiving    867413 non-null int64
ZipCode          867413 non-null object
PricePerSqFt     867413 non-null float64
dtypes: float64(3), int64(2), object(2)
memory usage: 52.9+ MB


### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [231]:
sales_price['DocumentDate']= pd.to_datetime(sales_price['DocumentDate'])
sales_price.head()

/Users/krex/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0  138860.0  110.0   2014-08-21     245000           1490   98002   
1  138860.0  110.0   1989-06-12     109300           1490   98002   
2  138860.0  110.0   2005-01-16      14684           1490   98002   
6  423943.0   50.0   1999-07-15      96000            960   98092   
7  423943.0   50.0   2001-01-08     127500            960   98092   

   PricePerSqFt  
0    164.429530  
1     73.355705  
2      9.855034  
6    100.000000  
7    132.812500

In [232]:
sales_2019 = sales_price[sales_price['DocumentDate'].apply(lambda x: x.year == 2019)]

sales_2019.info()

#sales_datak = sales_datak.loc[sales_datak['DocumentDate']. > 0, :]

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3626 entries, 605 to 1437220
Data columns (total 7 columns):
Major            3626 non-null float64
Minor            3626 non-null float64
DocumentDate     3626 non-null datetime64[ns]
SalePrice        3626 non-null int64
SqFtTotLiving    3626 non-null int64
ZipCode          3626 non-null object
PricePerSqFt     3626 non-null float64
dtypes: datetime64[ns](1), float64(3), int64(2), object(1)
memory usage: 226.6+ KB


### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [234]:
seattle_zips = [98101, 98102, 98103, 98104, 98105, 98106, 98107, 98108, 98109, 98112, 98115, 98116, 98117, 98118, 98119, 98121, 98122, 98125, 98126, 98133, 98134, 98136, 98144, 98146, 98154, 98164, 98174, 98177, 98178, 98195, 98199]

In [235]:
seattle_zip_strings= []
for zip in seattle_zips:
    seattle_zip_strings.append(str(zip))
seattle_zip_strings

['98101',
 '98102',
 '98103',
 '98104',
 '98105',
 '98106',
 '98107',
 '98108',
 '98109',
 '98112',
 '98115',
 '98116',
 '98117',
 '98118',
 '98119',
 '98121',
 '98122',
 '98125',
 '98126',
 '98133',
 '98134',
 '98136',
 '98144',
 '98146',
 '98154',
 '98164',
 '98174',
 '98177',
 '98178',
 '98195',
 '98199']

In [237]:
seattle_sales_2019 = sales_2019[sales_2019["ZipCode"].apply(lambda x: x.startswith("981"))]
seattle_sales_2019.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1495 entries, 651 to 1437220
Data columns (total 7 columns):
Major            1495 non-null float64
Minor            1495 non-null float64
DocumentDate     1495 non-null datetime64[ns]
SalePrice        1495 non-null int64
SqFtTotLiving    1495 non-null int64
ZipCode          1495 non-null object
PricePerSqFt     1495 non-null float64
dtypes: datetime64[ns](1), float64(3), int64(2), object(1)
memory usage: 93.4+ KB


### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [238]:
seattle_sales_2019['PricePerSqFt'].mean()

471.5043638771971